adapters/dkb-adapter/src/main/java/de/adorsys/xs2a/gateway/service/impl/DkbAccessTokenService.java

```diff
@@ -0,0 +1,150 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package de.adorsys.xs2a.gateway.service.impl;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import de.adorsys.xs2a.gateway.http.HttpClient;
+import de.adorsys.xs2a.gateway.http.JsonMapper;
+import de.adorsys.xs2a.gateway.security.AccessTokenException;
+import de.adorsys.xs2a.gateway.security.AccessTokenService;
+import de.adorsys.xs2a.gateway.service.GeneralResponse;
+
+import java.time.LocalDateTime;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.Map;
+
+public class DkbAccessTokenService implements AccessTokenService {
+
+    public static final String DKB_TOKEN_CONSUMER_KEY_PROPERTY = "dkb.token.consumer_key";
+    public static final String DKB_TOKEN_CONSUMER_SECRET_PROPERTY = "dkb.token.consumer_secret";
+    private static final String DKB_TOKEN_URL_PROPERTY = "dkb.token.url";
+    private static final String DKB_TOKEN_SECONDS_BEFORE_TOKEN_EXPIRATION_PROPERTY = "dkb.token.seconds_before_token_expiration";
+
+    private static final String DEFAULT_SECONDS_BEFORE_TOKEN_EXPIRATION = "60";
+    private static final String DEFAULT_TOKEN_URL = "https://api.dkb.de/token";
+
+    private static AccessTokenService instance;
+    private static Map<String, String> headers;
+    private static String tokenUrl;
+    private static int secondsBeforeTokenExpiration;
+
+    private HttpClient httpClient;
+
+    private AccessToken accessToken;
+
+    private JsonMapper jsonMapper;
+
+
+    private DkbAccessTokenService() {
+        jsonMapper = new JsonMapper();
+        setHttpClient(HttpClient.newHttpClient());
+    }
+
+    public static AccessTokenService getInstance() {
+        if (instance == null) {
+            instance = new DkbAccessTokenService();
+        }
+        return instance;
+    }
+
+    static {
+        String consumerKey = System.getProperty(DKB_TOKEN_CONSUMER_KEY_PROPERTY);
+        String consumerSecret = System.getProperty(DKB_TOKEN_CONSUMER_SECRET_PROPERTY);
+
+        if (consumerKey.isEmpty() || consumerSecret.isEmpty()) {
+            throw new AccessTokenException("Consumer key or secret are not provided");
+        }
+
+        headers = new HashMap<>();
+        headers.put("Content-type", "application/x-www-form-urlencoded");
+        headers.put("Authorization", "Basic " + buildBasicAuthorization(consumerKey, consumerSecret));
+
+        tokenUrl = System.getProperty(DKB_TOKEN_URL_PROPERTY, DEFAULT_TOKEN_URL);
+
+        secondsBeforeTokenExpiration = Integer.valueOf(System.getProperty(
+                DKB_TOKEN_SECONDS_BEFORE_TOKEN_EXPIRATION_PROPERTY,
+                DEFAULT_SECONDS_BEFORE_TOKEN_EXPIRATION
+        ));
+    }
+
+    @Override
+    public String retrieveToken() {
+        if (!isValid()) {
+            GeneralResponse<TokenResponse> response = httpClient.post(
+                    tokenUrl,
+                    "grant_type=client_credentials",
```

Comment:
Could you, please, extract this String to some constants to increase the code readability?

Reply:
Thanks, have moved to constants